### PR TITLE
Unified notifications

### PR DESCRIPTION
**Pull request by anvmn**

### CODE REVIEW

**Comment by anvmn** [comment]
@ordavidil 
I think that with 19 kinds of message types, Drupal installation takes longer than 10 minutes,
and Travis stops it.

**Comment by ordavidil** [comment]
@anvmn 
How long it takes on your machine ?

**Comment by anvmn** [comment]
@ordavidil 
Full installation, without migration - 24 minutes.

**Comment by ordavidil** [comment]
@anvmn 
Fixed ...

**Comment by soniCaH** [comment]
@anvmn when testing the first issue that was merged in (CFM-454), there seems to be some functionality broken:

membershipreq

**Comment by anvmn** [comment]
@soniCaH 
Checking...

**Comment by soniCaH** [comment]
- [ ] 454
- [ ] 478
- [ ] 474
- [ ] 475
- [ ] 476
- [ ] 474
- [ ] 477
- [ ] 481
- [ ] 874
- [ ] 961
- [ ] 962
- [ ] 472
- [ ] 1252
- [ ] 1220
- [ ] 869
- [ ] 945
- [ ] 889
- [ ] 836
- [ ] 1251

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 1]
```diff
@@ -1863,6 +1883,46 @@ function c4m_message_format_node_url(Message $message) {
...
+ * @return string
+ *   Direct path to comment.
+ */
+function c4m_message_format_comment_url(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+
+  $node_url = c4m_message_format_node_url($message);
+  $comment_id = $wrapper->field_comment->value(array('identifier' => TRUE));
+
+  return $node_url . '#comment-' . $comment_id;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
add a comment with an example of how it should looks like.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 2]
```diff
@@ -690,11 +691,18 @@ function _c4m_user_notifications_comment_on_content($cid, $nid) {
   $node = node_load($nid);
   // Flag suffix that is used at unsubscribe mechanism.
   $flag_suffix = 'comment';
+  // Additional flag, entity has (or has not) to be marked with,
+  // for user to be selected for notification.
+  // With 'flags_operation' set to 'diff', only user that got entity
+  // marked with 'follow comment' and not 'follow node' flags will be selected.
+  $additional_flag_suffix = 'node';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
this is not really clear ...

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 3]
```diff
@@ -781,11 +789,13 @@ function c4m_user_notifications_message_subscribe_get_subscribers($message, $sub
...
 
   // If we need to filter users according to additional flag.
-  if (isset($context['additional_flag_suffix'])) {
+  if (isset($context['additional_flag_suffix']) && isset($context['flags_operation'])) {
     // Filtering.
     $flag_name = 'subscribe_c4m_follow_' . $context['additional_flag_suffix'];
     $filtered_users = flag_get_entity_flags($flagged_entity_type, $entity_id, $flag_name);
-    $filtered_users = array_intersect(array_keys($users), array_keys($filtered_users));
+    // Flag operations can be 'intersect' or 'diff'.
+    $function = 'array_' . $context['flags_operation'];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
maybe we should be more explicit over here

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module (hunk 1 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn ^^

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 4]
```diff
@@ -244,4 +679,156 @@ function c4m_user_notifications_notify_inactive_users() {
...
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $wrapper->field_group_node->set($node->nid);
+  $wrapper->field_executing_user->set($user->uid);
+
+  // Message will be saved in message_notify_send_message().
+  message_notify_send_message($wrapper->value());
+}
+
+/**
+ * Returns an array of subscribers to notify.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: `Implements hook_message_subscribe_get_subscribers().`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 5]
```diff
@@ -244,4 +679,156 @@ function c4m_user_notifications_notify_inactive_users() {
...
+ *
+ * Function message_subscribe_get_subscribers() implementation, which is invoked
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: fix docs

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 6]
```diff
@@ -244,4 +679,156 @@ function c4m_user_notifications_notify_inactive_users() {
...
+  $filtered_users = array_keys($users);
+
+  // If we need to filter users according to additional flag.
+  if (isset($context['additional_flag_suffix']) && isset($context['flags_operation'])) {
+    // Filtering.
+    $flag_name = 'subscribe_c4m_follow_' . $context['additional_flag_suffix'];
+    $filtered_users = flag_get_entity_flags($flagged_entity_type, $entity_id, $flag_name);
+    // Flag operations can be 'intersect' or 'diff'.
+    $function = 'array_' . $context['flags_operation'];
+    $filtered_users = $function(array_keys($users), array_keys($filtered_users));
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: Add an elaborated comment over here

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 7]
```diff
@@ -170,113 +525,1600 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+ * @return string
+ *   Returns member's country.
+ */
+function c4m_message_format_new_member_country(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $field = field_info_field('c4m_country');
+  $allowed_values = list_allowed_values($field);
+  $index = $wrapper->field_related_user->c4m_country->value();
+
+  return empty($index) ? '' : $allowed_values[$index];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
How is it possible to get an empty index ?

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: In case c4m_country is not set.
This field is defined as required, but when using Admin user, it's not set.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: If you're talking about `user 1` then I'm not sure it should get this kind of notifications

@soniCaH ,
I'm not sure if this was discussed before, but, should this user get any of these notifications ?

Comment (soniCaH) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: `c4m_country` can be empty for migrated users as this field was not mandatory in the old platform.

Regarding user 1:
Only if he legitimately should receive them (not because he is user "1"): so only if he is group owner, manager...

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 8]
```diff
@@ -96,7 +96,8 @@ function _c4m_message_format_welcome_to_group_from_field(&$message, $message_ent
 /**
  * Helper function.
  *
- * Formats 'c4m_group_admin_demoted' message From field, as defined at template.
+ * Formats 'c4m_notify_admins_membership_approved' message From field,
+ * as defined at template.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
Please provide the template ID

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 9]
```diff
@@ -500,7 +500,7 @@ function c4m_user_notifications_notify_admins_admin_granted($gid, $uid) {
 }
 
 /**
- * Email notification to user about user him being revoked admin status.
+ * Email notification to user about user being revoked admin status.
  */
 function c4m_user_notifications_notify_user_admin_demoted($gid, $uid) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
Shouldn't this be `group_admin_demoted` ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 10]
```diff
@@ -517,12 +517,14 @@ function c4m_user_notifications_notify_user_admin_demoted($gid, $uid) {
  *   User id.
  */
 function _c4m_user_notify_user_admin_demoted($gid, $uid) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
Shouldn't this be `group_admin_demoted` ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 9 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: Also, why do we need to call this one and not just calling `_c4m_user_notify_user_admin_demoted()` directly ?

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 11]
```diff
@@ -345,11 +345,12 @@ function c4m_og_control_membership($action, $gid, $account, $membership, $timest
   if ($membership->state == OG_STATE_PENDING) {
     $membership->state = $action == 'approve' ? OG_STATE_ACTIVE : OG_STATE_BLOCKED;
     og_membership_save($membership);
-    $message = $action == 'approve' ? t('Membership approved!') : t('Membership rejected!');
+    $message = $action == 'approve' ? t('Membership approved') : t('Membership rejected');
     drupal_set_message($message);
   }
   else {
-    drupal_set_message(t('Membership request already handled!'), 'warning');
+    $message = $membership->state == OG_STATE_ACTIVE ? t('This request was already approved') : t('This request was already rejected');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: @anvmn 
why did you change the sentence ?

plus,
let's be explicit with the condition
```php
if ($membership->state == OG_STATE_ACTIVE) { ...
```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 10 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: We have 2 functions for this event:
c4m_user_notifications_notify_user_admin_demoted - notify user that was demoted
c4m_user_notifications_notify_admins_admin_demoted - notify GO/GA that admin was demoted

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 9 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: Moved `_c4m_user_notify_user_admin_demoted()`  logic into 
c4m_user_notifications_notify_user_admin_demoted().

**Comment by soniCaH** [comment]
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-454
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-481
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-869
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-889
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-945
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-962
- [x] https://issuetracker.amplexor.com/jira/browse/CFM-1252
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-472
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-474
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-475
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-476
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-477
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-478
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-836
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-874
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-961 - No notifications sent
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-1220
- [ ] https://issuetracker.amplexor.com/jira/browse/CFM-1251

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 12]
```diff
@@ -1949,7 +1949,9 @@ function c4m_message_format_node_title(Message $message) {
 function c4m_message_format_node_type(Message $message) {
   $wrapper = entity_metadata_wrapper('message', $message);
 
-  return $wrapper->field_node->value()->type;
+  $node = $wrapper->field_node->value();
+
+  return node_type_get_name($node);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: `return $node->type;`
should be enough here ...

Comment (soniCaH) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: You get "wiki_page" then for example.

```Dear Jean-Claude Juncker,

Albert Einstein added a new comment on the following wiki_page: Nominations```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 13]
```diff
@@ -557,37 +557,28 @@ function c4m_user_notifications_existing_node_updated($node) {
...
-function _c4m_user_notifications_content_recommended($nid) {
+function c4m_user_notifications_content_recommended($flag, $entity_id, $account, $flagging) {
   global $user;
   $message = message_create('c4m_notify_content_recommendation');
   // Define functions with which to resolve email template tokens.
   c4m_message_load_tokens__c4m_notify_content_recommendation($message);
 
   $wrapper = entity_metadata_wrapper('message', $message);
-  $wrapper->field_node->set($nid);
+  $wrapper->field_node->set($entity_id);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
if we set a node, why not leaving the `$nid` ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 14]
```diff
@@ -293,18 +318,45 @@ function c4m_user_notifications_form_og_ui_add_users_alter(&$form, &$form_state)
...
+
+  if ($flag->name != 'c4m_og_content_recommend') {
     // If not recommendation flag, return early.
     return;
   }
 
-  c4m_user_notifications_notify_author_content_recommended($flag, $entity_id, $account, $flagging);
-  c4m_user_notifications_notify_subscribed_editor_content_recommended($flag, $entity_id, $account, $flagging);
+  $node = node_load($entity_id);
+  $notification_types = array('wiki_page', 'discussion', 'document', 'event');
```
Comment (soniCaH) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn Do you know if this is somewhere in the specs? Couldn't find it somewhere myself.

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @soniCaH 
Well, actually, this is the info I got from you:
selection_061

Comment (soniCaH) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: Ok thanks.
Do we need to check this on multiple places, or is this the only place? Otherwise it might be worth abstracting this so this list can be managed in 1 place?

**Comment by soniCaH** [comment]
screen shot 2016-09-07 at 16 47 13

Go a weird activity stream message when adding a wiki page, with the "send notification" checkbox UNCHECKED.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module (hunk 14 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @soniCaH 
There're several places this info is used, and we have _c4m_user_notifications_get_group_content_notification_types() function that is used to get this info.
Also, I see we have function c4m_og_get_group_content_notification_types() at c4m_og module, that returns same info.

**Comment by ordavidil** [comment]
@soniCaH 
since we didn't touch the activity stream at all on this PR, I'm not sure what's the reason.
Also, as I've checked, the body of this message _c4m_insert__node__wiki_page_ was not changed, nor the functions that render the content of it ...

Are you on a clean installation ?
Did you checkout between branches while checking this PR ?
Is it possible that the configuration has been changed at some point ?

Anyway, can you please elaborate about the exact steps you took in order to get it?

Thanks

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module (hunk 14 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn 
in that case, let's remove one of them and just use the other one wherever we need.

**Comment by anvmn** [comment]
@ordavidil, @soniCaH 
Activity stream should know to present log messages, and not to present mail messages.
We have the filtering at:
`function kapablo_preprocess_message(&$variables) {
  if ($variables['view_mode'] != 'activity_group' && $variables['view_mode'] != 'activity_global') {
    return;
  }

  $message = $variables['message'];
  if (strpos($message->type, 'c4m_insert__') !== 0) {
    // Activity stream should include only log messages about new content -
    // who's name starts with 'c4m_insert__'.
    // We have another type - email message notifications. Here, we sort out
    // email messages, which shouldn't be processed for activity stream.
    return;
  }`

File: project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc [hunk 15]
```diff
@@ -16,6 +16,14 @@ function kapablo_preprocess_message(&$variables) {
   }
 
   $message = $variables['message'];
+  if (strpos($message->type, 'c4m_insert__') !== 0) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc]: @anvmn 
It seems to cause problems,
please double check this one

**Comment by soniCaH** [comment]
FYI; I did start from a clean install without changing branches in-between. Only thing I did was a pull to get latest commits yesterday.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 16]
```diff
@@ -350,7 +350,14 @@ function c4m_user_notifications_flag_flag($flag, $entity_id, $account, $flagging
   }
 
   $node = node_load($entity_id);
-  $notification_types = array('wiki_page', 'discussion', 'document', 'event');
+  $notification_types = array(
+    'article',
```
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @soniCaH,
For notification for articles and news, need to make sure that users are subscribed to relevant flag.

File: project/profiles/capacity4more/modules/c4m/content/c4m_content/c4m_content.module [hunk 17]
```diff
@@ -67,6 +67,23 @@ function c4m_content_form_node_form_alter(&$form, &$form_state, $form_id) {
 }
 
 /**
+ * Implements hook_node_submit().
+ *
+ * If 'Send Notification' checkbox exists, add a parameter with value to node.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/content/c4m_content/c4m_content.module]: `... to the node in order to determine if we should send the notification on the hook_node_update / insert`

File: project/profiles/capacity4more/modules/c4m/content/c4m_content/c4m_content.module [hunk 18]
```diff
@@ -14,14 +14,14 @@
 function c4m_content_form_node_form_alter(&$form, &$form_state, $form_id) {
   // Checks if the form's content type is being used as a message,
   // to add a send notification option.
-  if (in_array($form_state['node']->type, c4m_message_valid_node_types())) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/content/c4m_content/c4m_content.module]: @anvmn 
if `c4m_message_valid_node_types()` was created especially for here and is not being used anywhere else, please remove it.

**Comment by soniCaH** [comment]
PR Merged.
https://issuetracker.amplexor.com/jira/browse/CFM-1220 to be fixed in separate PR.
